File: orca-run-multiple-pipelines/src/test/java/io/armory/plugin/smp/RunMultiplePipelinesFixture.java
package io.armory.plugin.smp;

import com.netflix.spinnaker.kork.plugins.internal.PluginJar;
import com.netflix.spinnaker.orca.StageResolver;
import com.netflix.spinnaker.orca.api.test.OrcaFixture;
import com.netflix.spinnaker.orca.front50.DependentPipelineStarter;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.TestPropertySource;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Comparator;

@ContextConfiguration(classes = DependentPipelineStarter.class)
@TestPropertySource(properties = {
  "spinnaker.extensibility.plugins.Armory.RunMultiplePipelines.enabled=true",
  "spinnaker.extensibility.plugins-root-path=build/plugins"
})
@AutoConfigureMockMvc
public class RunMultiplePipelinesFixture extends OrcaFixture {

    @Autowired
    StageResolver stageResolver;

    private static File pluginsFolder;

    @BeforeAll
    public static void setup() {
        String pluginId = "Armory.RunMultiplePipelines";

        pluginsFolder = new File("build/plugins");
        pluginsFolder.mkdirs();

        new PluginJar.Builder(pluginsFolder.toPath().resolve(pluginId + ".jar"), pluginId)
                .pluginClass(RunMultiplePipelinesPlugin.class.getName())
                .pluginVersion("1.0.0")
                .build();
    }

    @AfterAll
    public static void cleanUp() throws IOException {
        Files.walk(pluginsFolder.toPath())
                .sorted(Comparator.reverseOrder())
                .map(Path::toFile)
                .forEach(File::delete);
    }
}

File: orca-run-multiple-pipelines/src/test/kotlin/io/armory/plugin/smp/RunMultiplePipelinesFixture.kt
/*
 * Copyright 2020 Armory, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.armory.plugin.smp

import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.netflix.spinnaker.kork.plugins.internal.PluginJar
import com.netflix.spinnaker.orca.StageResolver
import com.netflix.spinnaker.orca.api.test.OrcaFixture
import com.netflix.spinnaker.orca.front50.DependentPipelineStarter
import com.netflix.spinnaker.orca.front50.Front50Service
import io.armory.plugin.smp.tasks.RunMultiplePipelinesTask
import java.io.File
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.test.context.ContextConfiguration
import org.springframework.test.context.TestPropertySource
import org.springframework.test.web.servlet.MockMvc

@ContextConfiguration(classes = [DependentPipelineStarter::class])
@TestPropertySource(properties = [
    "spinnaker.extensibility.plugins.Armory.RunMultiplePipelinesPlugin.enabled=true",
    "spinnaker.extensibility.plugins-root-path=build/plugins",
    "front50.enabled=true",
    "front50.base-url=http://localhost:8084/"
])
@AutoConfigureMockMvc
class OrcaPluginsFixture : OrcaFixture() {

    @Autowired
    lateinit var stageResolver: StageResolver

    @Autowired
    lateinit var mockMvc: MockMvc

    val mapper = jacksonObjectMapper().disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)

    init {
        val pluginId = "Armory.RunMultiplePipelinesPlugin"
        val plugins = File("build/plugins").also {
            it.delete()
            it.mkdir()
        }

        PluginJar.Builder(plugins.toPath().resolve("$pluginId.jar"), pluginId)
            .pluginClass(RunMultiplePipelinesPlugin::class.java.name)
            .pluginVersion("0.0.1")
            .extensions(mutableListOf(RunMultiplePipelinesStage::class.java.name, RunMultiplePipelinesTask::class.java.name))
            .build()
    }
}

